include/bout/hypre_interface.hxx

```diff
@@ -810,11 +821,26 @@ public:
 
     comm = std::is_same<T, FieldPerp>::value ? mesh.getXcomm() : BoutComm::get();
 
-    HYPRE_ParCSRGMRESCreate(comm, &solver);
-    /* set the GMRES parameters */
-    //HYPRE_GMRESSetKDim(solver, 5); // commented out to let Hypre pick default
-    HYPRE_GMRESSetMaxIter(solver, 200);
-    HYPRE_GMRESSetTol(solver, 1.0e-7);
+    auto print_level = options["hypre_print_level"]
+                         .doc("Verbosity for Hypre solver. Integer from 0 (silent) to 4 "
+                              "(most verbose).")
+                         .withDefault(0);
+
+    if (solver_type == HYPRE_SOLVER_TYPE::gmres) {
+      HYPRE_ParCSRGMRESCreate(comm, &solver);
+      /* set the GMRES parameters */
+      //HYPRE_GMRESSetKDim(solver, 5); // commented out to let Hypre pick default
+      HYPRE_ParCSRGMRESSetPrintLevel(solver, print_level);
+    } else {
+      throw BoutException("Unsupported hypre_solver_type {}", solver_type);
+    }
```

Comment:
I usually prefer flipping the conditionals for preconditions to exit early, which leaves the rest of the code less-indented:

```suggestion
    if (solver_type != HYPRE_SOLVER_TYPE::gmres) {
      throw BoutException("Unsupported hypre_solver_type {}", solver_type);
    }
    HYPRE_ParCSRGMRESCreate(comm, &solver);
    /* set the GMRES parameters */
    //HYPRE_GMRESSetKDim(solver, 5); // commented out to let Hypre pick default
    HYPRE_ParCSRGMRESSetPrintLevel(solver, print_level);
```

I think you could also get away with removing the checks in the member functions, as they would already need an instance that was created with an acceptable `solver_type`

Reply:
The intention is to have some other solver types too (CG and BiCGStab maybe?) that need different HYPRE functions calling, so this should become an `if...else if...else if...else`. Here I was just setting up the option so Holger/Aaron can add the extra cases.

I tend to put the `else {throw ...}` everywhere just in case one of the places is forgotten when implementing a new `solver_type`. It's true they don't catch user errors, so don't matter as long as the developers are careful, so I'm happy to remove the checks in member functions if you prefer @ZedThree.

Reply:
ah yes, actually looking at the Hypre function names it's more obvious now! In that case, maybe consider a `switch` instead? Missing off the `default` case and not throwing is a possibility -- the compiler should warn about any missing enum values, and `Options` should have thrown already if it's not an acceptable value.

This isn't a blocker, just a suggestion :)